### DISCUSSION TITLE
Generic type argument inference fails in the face of implicit cast (e.g. Span<T> to ReadOnlySpan<T>)

### DISCUSSION THREAD

**Discussion by nietras** [dotnet/csharplang#6823] [General]
**Version Used**:  Compiler version: '4.5.0-2.22580.13 (ea0b6351)'. Language version: 11.0.

Even a simple program like below using `Span<T>` will fail in the face of implicit cast. This means we basically have to have both `Span<T>` and `ReadOnlySpan<T>` overloads for **all** methods relevant for this which is a huge hassle. This of course extends to other cases not just `Span<>` but it has become quite pervasive with the introduction of spans and read only spans, since implicit casts are everywhere now.

```csharp
Span<int> ints = stackalloc int[2];

// Error CS0411 The type arguments for method 'Method<T>(ReadOnlySpan<T>)'
// cannot be inferred from the usage. Try specifying the type arguments
// explicitly.
Method(ints);

static void Method<T>(ReadOnlySpan<T> span) { }
```

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

</Project>
```

**Diagnostic Id**: `CS0411`

**Expected Behavior**: Infer the type! 🙏 Especially when the implicit type conversion is clear (e.g. there is just one or something).

**Actual Behavior**: Compile error.

PS: This probably has been reported before, but I could not find any existing issue probably due to lack of search skills. Or it might have to be reported on `csharplang`, I don't know. 😅

**Comment by CyrusNajmabadi** [comment]
This looks to be a language request.
